PocketTutor-Team/pockettutor-app
# Feat/tutor match

# Add a very simple matching between student's lessons and tutor's profiles

### PR Description
This PR complete the life cycle of the lesson by implementing the matching algorithm part. For now it is a very simple algorithm that simply filter tutors based on the subject they teach and the price they want. We will need to implement a more flexible matching between the timeslots of the lesson and the availability of the tutor.

So, when the user is a student and he create a lesson (using the "Find a tutor" button), after filling all the needed fields, he will be directed to a new screen (TutorMatchingScreen in TutorMatching.kt) where he will see all the tutors that corresponds to his request (Screenshot 1). He can then click on a tutor and a pop-up will appear to confirm he want to have his lesson with this tutor (Screenshot 2). If the student chooses a tutor, it will appears in his overview in the section "Waiting for Tutor Confirmation" (STUDENT_REQUESTED state of the lesson) and from the point of view of the tutor, it will appear in the "Matched Student Lessons" section.

In order for the lesson cycle to be complete, we also need to be able (both from tutor and student side) to confirm a lesson so that we are sure that evrything is noted from both parts when a lesson is ready to be given, This is also done in this PR. To do so, I re-used screens (TutorMatchingScreen for student confirmation and TutorLessonResponseScreen for tutor confirmation) and updated them so that depending on the current state of the lesson it updates correctly the new state.

#### Modifed files
- TutorMatching.kt => screen that displays the list of matched tutors and allow the selection of one by the student. This screen is also used by the stident side to confirm a lesson 
- DisplayTutors.kt => composable that displays the given list of tutors (~DisplayLessons.kt)

- TutorLessonResponse.kt => update this screen so that it can also be used for tutor confirmation of a lesson

- AddLesson.kt => remove the save of the lesson in the database and navigate to the TutorMatchingScreen instead (which will save it in the database)

- MainActivity.kt and NavigationActions.kt => update the navigation to be able to navigate to the new screens and to the confirmation screen properly
- Overview.kt => update the section displayed on tutor and student overview to be compatible with the lesson cycle and update the function that allow us to click on a lesson either to modifiy it (only for student) or to confirm it (depending on its state)

- LessonStatus.kt and DisplayLessons.kt => remove the REQUESTED state that is now replaced by STUDENT_REQUESTED and TUTOR_REQUESTED

- LessonRepository.kt, LessonRepositoryFirestore.kt and LessonViewModel.kt=> rename the getRequestedLesson method to a more intuitive name given that the status we want is PENDING 

- LessonEditor.kt => fix the status of the updated lesson to be the same as the given lesson if any
- RequestedLessonTest.kt => fix the status of the mock lesson with status that correspond to the (current) lesson life cycle 

- EditRequestedLesson.kt => remove the lessonId parameter which is unecessary because we have the selected lesson directly in the view model

#### How to test it
The more straightforward way to test it is to run the MainActivity. The new added screen will be seen when a student create a new lesson. To test the interaction between all states and between tutor and student, the more easy way is to use two phones with two different users, one tutor and one student, and create lessons together. Note that this was already done with Alexis Rufer before publishing the PR.

#### Screenshots
![image](https://github.com/user-attachments/assets/f1298190-fcb7-4c4a-81f4-293bb4233898) ![image](https://github.com/user-attachments/assets/090348b4-5740-4dd2-9ef5-3829e7438d6a)
